File: stubs/csharp/src/BeastMode.RmDdd.Examples/Examples/ECommerce/CustomerId.cs
using System;
using System.Collections.Generic;
using BeastMode.RmDdd.Domain;
using BeastMode.RmDdd.Utilities;

namespace BeastMode.RmDdd.Examples.ECommerce
{
    /// <summary>
    /// Value object representing a Customer identifier.
    /// </summary>
    public class CustomerId : ValueObject
    {
        public Guid Value { get; }
        
        public CustomerId(Guid value)
        {
            Value = value;
            ValidateOnConstruction();
        }
        
        public CustomerId(string value) : this(Guid.Parse(value))
        {
        }
        
        public static CustomerId Generate()
        {
            return new CustomerId(Guid.NewGuid());
        }
        
        public override ValidationResult Validate()
        {
            var result = new ValidationResult();
            
            if (Value == Guid.Empty)
            {
                result.AddError("Customer ID value cannot be empty");
            }
            
            return result;
        }
        
        protected override IEnumerable<object?> GetEqualityComponents()
        {
            yield return Value;
        }
        
        public override string ToString()
        {
            return $"CustomerId({Value})";
        }
        
        // Implicit conversion operators for convenience
        public static implicit operator Guid(CustomerId customerId) => customerId.Value;
        public static implicit operator CustomerId(Guid value) => new(value);
    }
}

File: stubs/java/src/main/java/com/beastmode/rmddd/examples/ecommerce/CustomerId.java
package com.beastmode.rmddd.examples.ecommerce;

import com.beastmode.rmddd.domain.ValueObject;
import com.beastmode.rmddd.utilities.ValidationResult;

import java.util.Objects;
import java.util.UUID;

/**
 * Value object representing a Customer identifier.
 */
public class CustomerId extends ValueObject {
    
    private final UUID value;
    
    public CustomerId(UUID value) {
        this.value = Objects.requireNonNull(value, "Customer ID value cannot be null");
        validateOnConstruction();
    }
    
    public CustomerId(String value) {
        this(UUID.fromString(value));
    }
    
    public static CustomerId generate() {
        return new CustomerId(UUID.randomUUID());
    }
    
    public UUID getValue() {
        return value;
    }
    
    @Override
    public ValidationResult validate() {
        ValidationResult result = new ValidationResult();
        
        if (value == null) {
            result.addError("Customer ID value cannot be null");
        }
        
        return result;
    }
    
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        CustomerId that = (CustomerId) o;
        return Objects.equals(value, that.value);
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(value);
    }
    
    @Override
    public String toString() {
        return "CustomerId{" + value + '}';
    }
}